### ISSUE TITLE
Versioning scheme and related policies

### ISSUE THREAD

**Issue by goodboy** [goodboy/tractor#116]
As per some discussion in #104 we'll figure out the "best" (read most hip) versioning scheme after we get an initial dev release on the ol' pypi.

Some notes from discussion I've had with @salotz:
- there's a [pep for dev releases](https://www.python.org/dev/peps/pep-0440/#developmental-releases)
- @salotz has a cool idea for a so called [semantic changelog](https://github.com/salotz/rfcs/blob/master/rfcs/salotz.002_semantic-changelog.org)
- the philosophy of [growth versioning](https://github.com/salotz/rfcs/blob/master/rfcs/salotz.010_growth-versioning.org)
- consideration for [codetags](https://github.com/salotz/rfcs/blob/master/rfcs/salotz.006_codetags.org)

**Comment by salotz** [comment]
Since then I've come up with some practical usages for the different tags. Ignoring alphas I make a series of `rc` releases to allow for usage and finding missed bugs. Then you can do `dev` releases in between to "debug" the actual packaging release process itself. I.e. if you screw up and forget to include necessary files in the dist.

For reference I keep a fairly up to date cookiecutter: https://github.com/salotz/salotz-py-cookiecutter/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D

Dev guide: https://github.com/salotz/salotz-py-cookiecutter/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/info/dev_guide.org#releases

Not sure the goal for this issue, but basically what is written there is what I would do. You can review in that repo if you want.

**Comment by goodboy** [comment]
> Not sure the goal for this issue, but basically what is written there is what I would do. You can review in that repo if you want.

@salotz mostly discussion and archiving all the above links :smiley_cat: 

I would like to automate the release process and change-log if possible with something like [towncrier](https://github.com/twisted/towncrier) or whatever is hip now.

**Comment by salotz** [comment]
Ok cool. You know me well to assign vaguely purposeful conversation :P 

re towncrier. I might suggest to just roll our own since basically their configuration is most of the same stuff you would configure in building a static site via e.g. Nikola. Which is quite easy to do.

That said rendering HTML is second priority to good practices of writing them, having a clean process, etc.

I usually have both a changelog for releases as well as a "news" feed. You can add semantics to the changelog potentially like towncrier leaving your news free of such shackles.

**Comment by goodboy** [comment]
Oh I was also going to mention that using [setuptools_scm](https://github.com/pypa/setuptools_scm) is probably a cool idea. 

We use it a bunch in the [`pytest` community projects](https://github.com/pytest-dev) and I've found it quite nice for avoiding managing versions by hand.

**Comment by goodboy** [comment]
Follow up discussion from #104.

- https://semver.org/#faq
- we could try [versioneer](https://github.com/python-versioneer/python-versioneer) for incremental automatic tagging from `git`
- @richardsheridan's [PR showing the modern dev's installation system setup](https://github.com/richardsheridan/trio-parallel/pull/28/commits/d8c8ca4a00d53ff04a80c7f38d949dea31d610c7) closely mirroring `trio`'s

**Comment by richardsheridan** [comment]
https://hynek.me/articles/semver-will-not-save-you/

**Comment by goodboy** [comment]
@richardsheridan yeah see calver has always appealed to me ever since seeing in salt-stack way back:
https://calver.org/

I personally prefer the time linking since it just gives the reader so much more info from the outset.
